Mathlib/GroupTheory/GroupAction/Blocks.lean

```diff
@@ -381,6 +395,163 @@ theorem IsBlockSystem.of_normal {N : Subgroup G} [N.Normal] :
 
 end Normal
 
+section Fintype
+
+theorem Setoid.nat_sum {α : Type _} [Finite α] {c : Set (Set α)} (hc : Setoid.IsPartition c) :
+    (finsum fun x : c => Set.ncard (x : Set α)) = Nat.card α := by
+  classical
+  have := Fintype.ofFinite α
+  simp only [finsum_eq_sum_of_fintype, Nat.card_eq_fintype_card, ← Set.Nat.card_coe_set_eq]
+  rw [← Fintype.card_sigma]
+  refine' Fintype.card_congr (Equiv.ofBijective (fun x => x.snd : (Σ a : ↥c, a) → α) _)
+  constructor
+  -- injectivity
+  rintro ⟨⟨x, hx⟩, ⟨a, ha : a ∈ x⟩⟩ ⟨⟨y, hy⟩, ⟨b, hb : b ∈ y⟩⟩ hab
+  dsimp at hab
+  rw [hab] at ha
+  rw [Sigma.subtype_ext_iff]
+  simp only [Subtype.mk_eq_mk, Subtype.coe_mk]
+  apply And.intro _ hab
+  exact ExistsUnique.unique (hc.2 b) ⟨hx, ha⟩ ⟨hy, hb⟩
+  -- surjectivity
+  intro a
+  obtain ⟨x, ⟨hx, ha : a ∈ x⟩, _⟩ := hc.2 a
+  use ⟨⟨x, hx⟩, ⟨a, ha⟩⟩
+
+theorem Set.ncard_coe {α : Type*} (s : Set α) :
+    s.ncard = Set.ncard (Set.univ : Set (Set.Elem s)) := by
+  apply Set.ncard_congr (fun a ha ↦ ⟨a, ha⟩)
+  · exact fun a ha ↦ by simp only [Set.mem_univ]
+  · simp [Subtype.mk_eq_mk]
+  · exact fun ⟨a, ha⟩ _ ↦ ⟨a, ha, rfl⟩
+
+/-- The cardinality of the ambient is the product of
+  of the cardinality of a block
+  by the cardinality of the set of translates of that block -/
+theorem IsBlock.ncard_block_mul_ncard_orbit_eq
+    [Finite X] [IsPretransitive G X] {B : Set X}
+    (hB : IsBlock G B) (hB_ne : B.Nonempty) :
+    Set.ncard B * Set.ncard (Set.range fun g : G => g • B) = Nat.card X := by
+  classical
+  have := Fintype.ofFinite X
+  rw [← Setoid.nat_sum (hB.isBlockSystem hB_ne).1]
+  simp only [finsum_eq_sum_of_fintype]
+  rw [Finset.sum_congr rfl]
+  · rw [Finset.sum_const, mul_comm]
+    congr
+    rw [← Set.ncard_coe_Finset, Finset.coe_univ, Set.ncard_coe]
+  · rintro ⟨x, ⟨g, rfl⟩⟩ _
+    exact Set.ncard_image_of_injective B (MulAction.injective g)
+
+/-- The cardinality of a block divides the cardinality of the ambient type -/
+theorem IsBlock.ncard_of_block_divides [Finite X] [IsPretransitive G X] {B : Set X}
+    (hB : IsBlock G B) (hB_ne : B.Nonempty) :
+    Set.ncard B ∣ Nat.card X :=
+  Dvd.intro _ (hB.ncard_block_mul_ncard_orbit_eq hB_ne)
+
+/-- A too large block is equal to ⊤ -/
+theorem is_top_of_large_block [hfX : Finite X] [hGX : IsPretransitive G X] {B : Set X}
```

Comment:
To enable dot notation:
```suggestion
theorem IsBlock.eq_top_of_large [hfX : Finite X] [hGX : IsPretransitive G X] {B : Set X}
```

Reply:
I am not sure I understood what I should do…